SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/convert/Clausifier.java

```diff
@@ -1226,40 +1219,123 @@ public SharedTerm getSharedTerm(final Term t, final boolean inCCTermBuilder, fin
 						} else if (fs.getName().equals("to_int")) {
 							addToIntAxioms(at, source);
 						} else if (fs.getName().equals("ite") && fs.getReturnSort() != mTheory.getBooleanSort()) {
-							pushOperation(new AddTermITEAxiom(res, source));
+							pushOperation(new AddTermITEAxiom(term, source));
 						} else if (fs.getName().equals("store")) {
 							addStoreAxiom(at, source);
 						} else if (fs.getName().equals("@diff")) {
 							addDiffAxiom(at, source);
+							mArrayTheory.notifyDiff((CCAppTerm) termInfo.getCCTerm());
 						}
 					}
-					if (needCCTerm(fs) && !inCCTermBuilder && at.getParameters().length > 0) {
-						final CCTermBuilder cc = new CCTermBuilder(source);
-						res.mCCterm = cc.convert(t);
-					}
 				}
 			}
-			if (t.getSort().isNumericSort()) {
-				boolean needsLA = t instanceof ConstantTerm;
-				if (t instanceof ApplicationTerm) {
-					final String func = ((ApplicationTerm) t).getFunction().getName();
+			if (term.getSort().isArraySort()) {
+				assert termInfo.getCCTerm() != null;
+				boolean isStore = false;
+				boolean isConst = false;
+				if (term instanceof ApplicationTerm) {
+					final ApplicationTerm at = (ApplicationTerm) term;
+					final String funcName = at.getFunction().getName();
+					isStore = funcName.equals("store");
+					isConst = funcName.equals("const");
+				}
+				mArrayTheory.notifyArray(termInfo.getCCTerm(), isStore, isConst);
+			}
+			if (term.getSort().isNumericSort()) {
+				boolean needsLA = term instanceof ConstantTerm;
+				if (term instanceof ApplicationTerm) {
+					final String func = ((ApplicationTerm) term).getFunction().getName();
 					if (func.equals("+") || func.equals("-") || func.equals("*") || func.equals("to_real")) {
 						needsLA = true;
 					}
 				}
 				if (needsLA) {
-					getLASolver().generateSharedVar(res, createMutableAffinTerm(new SMTAffineTerm(t), source));
-					addUnshareLA(res);
+					final MutableAffineTerm mat = createMutableAffinTerm(new SMTAffineTerm(term), source);
+					assert mat.getConstant().mEps == 0;
+					shareLATerm(term, termInfo, new LASharedTerm(term, mat.getSummands(), mat.getConstant().mReal));
 				}
 			}
 		}
+	}
+
+	public LinVar getLinVar(final Term term) {
```

Comment:
Method only for basic term, right?

Reply:
may be only called for "initiallynonbasic term", i.e. a term that can appear in the summand hashmap of an SMTAffineTerm. I can add an assertion like
`term == new SMTAffineTerm(term).getSummands().keySet().iterator().next()`

Currently this is only used by model production, to get the value for each summand for a congruence-class that has a shared term.